### ISSUE TITLE
Offset is outside the bounds of the DataView

### ISSUE THREAD

**Issue by slifty** [gliese1337/HLS.js#14]
Piping an ffmpeg-generated mpegts stream and after a few minutes I occasionally get the following error:

```
/Users/slifty/Dropbox/Code/NodeJS/tvkitchen/appliances/node_modules/ts-demuxer/dist/index.js:402
  	throw e
  	^

RangeError: Offset is outside the bounds of the DataView
    at DataView.getUint8 (<anonymous>)
    at demux_packet (/Users/slifty/Dropbox/Code/NodeJS/tvkitchen/appliances/node_modules/ts-demuxer/dist/index.js:398:13)
    at TSDemuxer.process (/Users/slifty/Dropbox/Code/NodeJS/tvkitchen/appliances/node_modules/ts-demuxer/dist/index.js:477:23)
    at Transform._transform (/Users/slifty/Dropbox/Code/NodeJS/tvkitchen/appliances/packages/core/lib/AbstractVideoIngestionAppliance.js:82:26)
    at Transform._read (_stream_transform.js:191:10)
    at Transform._write (_stream_transform.js:179:12)
    at doWrite (_stream_writable.js:454:12)
    at writeOrBuffer (_stream_writable.js:436:5)
    at Transform.Writable.write (_stream_writable.js:327:11)
    at Socket.ondata (_stream_readable.js:708:22)
error Command failed with exit code 1.
```

I did some debugging and it turns out that ptr has a negative value (e.g. `-112`), so it makes sense that is out of bounds.  I haven't yet figured out what about the data leads ptr to that state but I will spend some time on that.

More broadly, though, I know that you're not actively working on this project.  If I do come up with a patch are you in a position to integrate it or should I look into forking and releasing a new package?

**Comment by gliese1337** [comment]
Yes, I can certainly update the module if you submit a patch or PR. Thanks!

**Comment by slifty** [comment]
Awesome -- I'll see if I can figure out what's up and get back to you!

**Comment by slifty** [comment]
I think I figured out [what's going wrong](https://github.com/gliese1337/HLS.js/blob/d995e9cbb8d72f663cb75280ba058f63dc1d2aec/demuxer/src/index.ts#L504-L515).

```
    for (let ptr = offset;;ptr += PACKET_LEN) {
      const datalen = len - ptr;
      this.ptr = datalen;
      if (datalen === 0) return 0; // complete packet
      if (datalen < PACKET_LEN) {
        this.leftover.set(buffer.subarray(ptr, ptr + datalen));
        return 1; // incomplete packet
      }

      const n = demux_packet(pmt, mem, ptr, pids, cb, false);
      if (n) return n // invalid packet
    }
```

The line `this.ptr = datalen` can set `this.ptr` to be larger than `PACKET_LEN` (since `datalen` is how much unprocessed data remains *in the buffer* and the buffer could be arbitrarily long / far longer than `PACKET_LEN`)

Then this line `if (n) return n // invalid packet` could potentially abort the processing at any point in the stream.

This means, for instance, that if buffer was 500 long and there was an invalid packet triggered in the first iteration of this loop, `datalen` (and `this.ptr`) would be 500.

Then, next time `process` is invoked we hit [this line](https://github.com/gliese1337/HLS.js/blob/d995e9cbb8d72f663cb75280ba058f63dc1d2aec/demuxer/src/index.ts#L484):

```
const remainder = (PACKET_LEN - this.ptr) % PACKET_LEN;
```

This sets `remainder = (188 - 500) % 188` which is `-124`

And from there we [sets offset](https://github.com/gliese1337/HLS.js/blob/d995e9cbb8d72f663cb75280ba058f63dc1d2aec/demuxer/src/index.ts#L500) to be `-124`:

```
offset += remainder;
```

And suddenly when we enter the for loop again `ptr` starts off negative (which blows up the moment we try to use `ptr` to access data).

---

I think that the culprit is `if (n) return n // invalid packet` since it completely aborts before cleaning up *any* state (e.g. it doesn't update `this.leftover` or `this.ptr`

**Comment by gliese1337** [comment]
OK, assuming this is the source of the problem, I would classify it more as a "feature request" than as a "bug", as that sequence of events can only occur if the client code receives an error and then tries to soldier on anyway--which this code was not designed to support!

However, being able to continue without having to throw everything away and restart is a reasonable thing to want to do. To that end, I have just pushed an update (https://github.com/gliese1337/HLS.js/commit/7152e09f0ca4743a589409ff5bb61b53344c6593) which will set `this.ptr` to zero (thus indicating no cached data) and will set a public `offset` property to notify the client code of where processing stopped in the input buffer. I have also added a static `resync(buffer, offset)` method to help with finding where to restart, and updated the README.
